src/init.cpp

```diff
@@ -170,12 +171,26 @@ static fs::path GetPidFile(const ArgsManager& args)
 #else
         tfm::format(file, "%d\n", getpid());
 #endif
+        g_generated_pid = true;
         return true;
     } else {
         return InitError(strprintf(_("Unable to create the PID file '%s': %s"), fs::PathToString(GetPidFile(args)), SysErrorString(errno)));
     }
 }
 
+static void RemovePidFile(const ArgsManager& args)
```

Comment:
#28784 stores the path in `static std::optional<std::string> g_generated_cookie`. I don't have a particularly strong conviction in either direction, but I'd use the same approach for both. Storing the location of the file to be deleted seems a bit more robust, so I think I'd lean towards this approach:

<details>
<summary>git diff on 98ba5ba0cc</summary>

```diff
diff --git a/src/init.cpp b/src/init.cpp
index afa0ab1ee6..f2d74db13f 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -96,6 +96,7 @@
 #include <cstdio>
 #include <fstream>
 #include <functional>
+#include <optional>
 #include <set>
 #include <string>
 #include <thread>
@@ -155,7 +156,7 @@ static const char* DEFAULT_ASMAP_FILENAME="ip_asn.map";
  * The PID file facilities.
  */
 static const char* BITCOIN_PID_FILENAME = "bitcoind.pid";
-static bool g_generated_pid{false};
+static std::optional<fs::path> g_generated_pid;
 
 static fs::path GetPidFile(const ArgsManager& args)
 {
@@ -164,25 +165,26 @@ static fs::path GetPidFile(const ArgsManager& args)
 
 [[nodiscard]] static bool CreatePidFile(const ArgsManager& args)
 {
-    std::ofstream file{GetPidFile(args)};
+    const fs::path pid_path{GetPidFile(args)};
+    std::ofstream file{pid_path};
     if (file) {
 #ifdef WIN32
         tfm::format(file, "%d\n", GetCurrentProcessId());
 #else
         tfm::format(file, "%d\n", getpid());
 #endif
-        g_generated_pid = true;
+        g_generated_pid = pid_path;
         return true;
     } else {
-        return InitError(strprintf(_("Unable to create the PID file '%s': %s"), fs::PathToString(GetPidFile(args)), SysErrorString(errno)));
+        return InitError(strprintf(_("Unable to create the PID file '%s': %s"), fs::PathToString(pid_path), SysErrorString(errno)));
     }
 }
 
-static void RemovePidFile(const ArgsManager& args)
+static void RemovePidFile()
 {
     if (!g_generated_pid) return;
     try {
-        if (!fs::remove(GetPidFile(args))) {
+        if (!fs::remove(g_generated_pid.value())) {
             LogPrintf("Unable to remove PID file: File does not exist\n");
         }
     } catch (const fs::filesystem_error& e) {
@@ -364,7 +366,7 @@ void Shutdown(NodeContext& node)
     node.scheduler.reset();
     node.kernel.reset();
 
-    RemovePidFile(*node.args);
+    RemovePidFile();
 
     LogPrintf("%s: done\n", __func__);
 }

```
</details>

Reply:
Oh, looking at 28784 more closely it seems `g_generated_cookie` stores the cookie, not the path. My bad.

Reply:
I prefer the idea of storing a bool with whether it has been generated or not, and use the `{Get|Create|Remove}PidFile()` functions as necessary.

Reply:
I'm okay with the boolean approach if you prefer it. I just like the "we created **this** file, so let's remove **this** file" explicitness over "we created **a** file, so let's remove **a** file". But in practice, no big difference.